src/main/java/it/pagopa/pm/gateway/service/async/VposAsyncService.java

```diff
@@ -0,0 +1,156 @@
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Service;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import static it.pagopa.pm.gateway.constant.Messages.GENERIC_ERROR_MSG;
+import static it.pagopa.pm.gateway.constant.VposConstant.*;
+import static it.pagopa.pm.gateway.dto.enums.PaymentRequestStatusEnum.*;
+
+@Service
+@Slf4j
+@NoArgsConstructor
+public class VposAsyncService {
+
+    private static final String CAUSE = " cause: ";
+    private static final List<String> METHOD_CHALLENGE_CODES = Arrays.asList("25", "26");
+    private String vposUrl;
+    private PaymentRequestRepository paymentRequestRepository;
+    private VPosRequestUtils vPosRequestUtils;
+    private VPosResponseUtils vPosResponseUtils;
+    private HttpClient httpClient;
+    private EcommercePatchUtils ecommercePatchUtils;
+
+    @Autowired
+    public VposAsyncService(@Value("${vpos.requestUrl}") String vposUrl, PaymentRequestRepository paymentRequestRepository,
+                            VPosRequestUtils vPosRequestUtils, VPosResponseUtils vPosResponseUtils,
+                            HttpClient httpClient, EcommercePatchUtils ecommercePatchUtils) {
+        this.vposUrl = vposUrl;
+        this.paymentRequestRepository = paymentRequestRepository;
+        this.vPosRequestUtils = vPosRequestUtils;
+        this.vPosResponseUtils = vPosResponseUtils;
+        this.httpClient = httpClient;
+        this.ecommercePatchUtils = ecommercePatchUtils;
+    }
+
+    @Async
+    public void executeStepZeroAuth(Map<String, String> params, PaymentRequestEntity entity, StepZeroRequest pgsRequest) {
+        try {
+            String requestId = entity.getGuid();
+            log.info("Calling VPOS - Step 0 - for requestId: " + requestId);
+            HttpClientResponse clientResponse = httpClient.callVPos(vposUrl,params);
+            ThreeDS2Response response = vPosResponseUtils.build3ds2Response(clientResponse.getEntity());
+            vPosResponseUtils.validateResponseMac(response.getTimestamp(), response.getResultCode(), response.getResultMac(), pgsRequest);
+            log.info("Result code from VPOS - Step 0 - for RequestId {} is {}", requestId, response.getResultCode());
+            boolean toAccount = checkResultCode(response, entity);
+            if (toAccount) {
+                String authNumber = ((ThreeDS2Authorization) response.getThreeDS2ResponseElement()).getAuthorizationNumber();
+                executeAccount(entity, pgsRequest, authNumber);
+            }
+
+            //If the resultCode is 25 or 26, the PATCH is not called
+            if (!METHOD_CHALLENGE_CODES.contains(response.getResultCode())) {
+                ecommercePatchUtils.executePatchTransaction(entity);
+            }
+        } catch (Exception e) {
+            log.error(GENERIC_ERROR_MSG + entity.getIdTransaction() + CAUSE + e.getCause() + " - " + e.getMessage(), e);
```

Comment:
```suggestion
            log.error("{}{}{} - {}", GENERIC_ERROR_MSG , entity.getIdTransaction() , CAUSE , e.getCause() , " - " , e.getMessage(), e);
```

Reply:
The suggestion is missing some placeholders.
Committed a corrected version on branch.